src/model_lifecycle.h

```diff
@@ -0,0 +1,283 @@
+
+  // Instruct the model to stop accepting new inference requests.
+  Status StopAllModels();
+
+  // Return the number of in-flight inference if any, model versions
+  // that don't have in-flight inferences will not be included.
+  const std::set<std::tuple<std::string, int64_t, size_t>> InflightStatus();
+
+ private:
+  struct ModelInfo {
+    ModelInfo(
+        const std::string& model_path,
+        const inference::ModelConfig& model_config,
+        const uint64_t last_update_ns)
+        : model_config_(model_config), model_path_(model_path),
+          is_ensemble_(false), last_update_ns_(last_update_ns),
+          state_(ModelReadyState::UNKNOWN)
+    {
+#ifdef TRITON_ENABLE_ENSEMBLE
+      is_ensemble_ = (model_config.platform() == kEnsemblePlatform);
+#endif  // TRITON_ENABLE_ENSEMBLE
+    }
+
+    inference::ModelConfig model_config_;
+    std::string model_path_;
+    bool is_ensemble_;
+
+    std::recursive_mutex mtx_;
+    uint64_t last_update_ns_;
+    ModelReadyState state_;
+    std::string state_reason_;
+
+    std::shared_ptr<TritonRepoAgentModelList> agent_model_list_;
+    std::shared_ptr<Model> model_;
+  };
+
+  struct LoadTracker {
+    LoadTracker(
+        const size_t affected_version_cnt, const uint64_t last_update_ns)
+        : load_failed_(false), completed_version_cnt_(0),
+          affected_version_cnt_(affected_version_cnt),
+          last_update_ns_(last_update_ns)
+    {
+    }
+
+    std::mutex mtx_;
+
+    bool load_failed_;
+    std::string reason_;
+    size_t completed_version_cnt_;
+    size_t affected_version_cnt_;
+    std::map<int64_t, ModelInfo*> load_set_;
+    uint64_t last_update_ns_;
+  };
+
+  ModelLifeCycle(
+      const double min_compute_capability, InferenceServer* server,
+      const triton::common::BackendCmdlineConfigMap& backend_cmdline_config_map,
+      const triton::common::HostPolicyCmdlineConfigMap& host_policy_map,
+      const unsigned int model_load_thread_count)
+      : min_compute_capability_(min_compute_capability), server_(server),
+        cmdline_config_map_(backend_cmdline_config_map),
+        host_policy_map_(host_policy_map)
+  {
+    load_pool_.reset(
+        new triton::common::ThreadPool(std::max(1u, model_load_thread_count)));
+  }
+
+  void CreateModel(
+      const std::string& model_name, const int64_t version,
+      ModelInfo* model_info);
+  // Callback function template for model load.
+  // 'OnComplete' needs to be passed by value for now as there can be
+  // multiple versions to be loaded and each holds a copy of
+  // the 'OnComplete' callback.
+  void OnLoadComplete(
+      const std::string& model_name, const int64_t version,
+      ModelInfo* model_info, std::function<void(Status)> OnComplete,
+      std::shared_ptr<LoadTracker> load_tracker);
+
+  const double min_compute_capability_;
+
+  // Mutex for 'map_' and 'background_models_'
+  std::recursive_mutex map_mtx_;
+
+  using VersionMap = std::map<int64_t, std::unique_ptr<ModelInfo>>;
```

Comment:
Why did VersionMap [used to map](https://github.com/triton-inference-server/core/blob/6a984dd747df4b0cc1782f63169a16ef9eb1ea7c/src/model_repository_manager.cc#L589-L591) to a `std::pair<ModelInfo*, ModelInfo*>` and now it just maps to a single `ModelInfo*`?

Was this how background models was handled previously?